File: Sources/ClipboardDrawer/Pipeline/ClipPipeline.swift
import Foundation

enum ClipPipelineTrigger: String, Codable, Sendable {
    case postCapture
    case manual
    case scheduled
}

enum ClipPipelineContent: Sendable {
    case text(String)
    case richText(plainText: String?, payloadPath: String)
    case html(plainText: String?, payloadPath: String)
    case image(payloadPath: String)
    case file(url: URL)
    case media(url: URL?)

    init(clip: ClipItem) {
        switch clip.type {
        case .text:
            self = .text(clip.plainText ?? "")
        case .rtf:
            self = .richText(plainText: clip.plainText, payloadPath: clip.payloadPath ?? "")
        case .html:
            self = .html(plainText: clip.plainText, payloadPath: clip.payloadPath ?? "")
        case .media:
            if let documentURL = clip.documentURL {
                self = .media(url: documentURL)
            } else if let payloadPath = clip.payloadPath {
                self = .image(payloadPath: payloadPath)
            } else {
                self = .media(url: nil)
            }
        case .document:
            self = .file(url: clip.documentURL ?? URL(fileURLWithPath: clip.plainText ?? ""))
        }
    }
}

struct ClipPipelineSource: Sendable {
    var appName: String?
    var bundleIdentifier: String?
    var capturedAt: Date
}

struct ClipPipelineArtifact: Identifiable, Sendable {
    let id: UUID
    let kind: String
    let title: String
    let content: ClipPipelineContent
    let producerPluginID: String
    let metadata: [String: String]

    init(
        id: UUID = UUID(),
        kind: String,
        title: String,
        content: ClipPipelineContent,
        producerPluginID: String,
        metadata: [String: String] = [:]
    ) {
        self.id = id
        self.kind = kind
        self.title = title
        self.content = content
        self.producerPluginID = producerPluginID
        self.metadata = metadata
    }
}

enum ClipPipelineStageStatus: Sendable, Equatable {
    case skipped
    case succeeded
    case failed
}

struct ClipPipelineStageResult: Sendable {
    let pluginID: String
    let status: ClipPipelineStageStatus
    let startedAt: Date
    let endedAt: Date
    let message: String?
}

struct ClipPipelineEffects: OptionSet, Sendable {
    let rawValue: Int

    static let saveDerivedClip = ClipPipelineEffects(rawValue: 1 << 0)
    static let copyToPasteboard = ClipPipelineEffects(rawValue: 1 << 1)
    static let pasteImmediately = ClipPipelineEffects(rawValue: 1 << 2)
    static let showBanner = ClipPipelineEffects(rawValue: 1 << 3)
}

struct ClipPipelineContext: Sendable {
    let runID: UUID
    let trigger: ClipPipelineTrigger
    let originalClip: ClipItem
    var current: ClipPipelineContent
    var source: ClipPipelineSource
    var metadata: [String: String]
    var artifacts: [ClipPipelineArtifact]
    var stageResults: [ClipPipelineStageResult]
    var requestedEffects: ClipPipelineEffects

    init(
        runID: UUID = UUID(),
        trigger: ClipPipelineTrigger,
        originalClip: ClipItem,
        current: ClipPipelineContent? = nil,
        source: ClipPipelineSource,
        metadata: [String: String] = [:],
        artifacts: [ClipPipelineArtifact] = [],
        stageResults: [ClipPipelineStageResult] = [],
        requestedEffects: ClipPipelineEffects = []
    ) {
        self.runID = runID
        self.trigger = trigger
        self.originalClip = originalClip
        self.current = current ?? ClipPipelineContent(clip: originalClip)
        self.source = source
        self.metadata = metadata
        self.artifacts = artifacts
        self.stageResults = stageResults
        self.requestedEffects = requestedEffects
    }
}

protocol ClipPipelinePlugin: Sendable {
    var id: String { get }
    var name: String { get }

    func canProcess(_ context: ClipPipelineContext) async -> Bool
    func process(_ context: ClipPipelineContext) async throws -> ClipPipelineContext
}

File: windows/Clipwell.Win/src/Clipwell.Core/Pipeline/ClipPipeline.cs
using Clipwell.Core.Models;

namespace Clipwell.Core.Pipeline;

/// <summary>
/// Plugin pipeline model, ported from the macOS app (Pipeline/ClipPipeline.swift).
/// A pipeline run threads a context through an ordered list of plugins; each plugin
/// may transform the current content, emit artifacts (derived clips), and request effects.
/// </summary>
public enum ClipPipelineTrigger
{
    PostCapture,
    Manual,
    Scheduled,
}

public enum ClipPipelineStageStatus
{
    Skipped,
    Succeeded,
    Failed,
}

public sealed record ClipPipelineStageResult(
    string PluginId,
    ClipPipelineStageStatus Status,
    DateTimeOffset StartedAt,
    DateTimeOffset EndedAt,
    string? Message);

[Flags]
public enum ClipPipelineEffects
{
    None = 0,
    SaveDerivedClip = 1 << 0,
    CopyToClipboard = 1 << 1,
    PasteImmediately = 1 << 2,
    ShowBanner = 1 << 3,
}

/// <summary>A derived output of a plugin (e.g. OCR text, a translation) tied to the producing plugin.</summary>
public sealed record ClipPipelineArtifact(
    string Kind,
    string Title,
    string Text,
    string ProducerPluginId,
    IReadOnlyDictionary<string, string>? Metadata = null);

public sealed class ClipPipelineContext
{
    public Guid RunId { get; } = Guid.NewGuid();
    public ClipPipelineTrigger Trigger { get; }
    public ClipItem OriginalClip { get; }

    /// <summary>Working text content, updated as plugins transform it. Null for image clips until OCR runs.</summary>
    public string? CurrentText { get; set; }

    public Dictionary<string, string> Metadata { get; } = [];
    public List<ClipPipelineArtifact> Artifacts { get; } = [];
    public List<ClipPipelineStageResult> StageResults { get; } = [];
    public ClipPipelineEffects RequestedEffects { get; set; } = ClipPipelineEffects.None;

    public ClipPipelineContext(ClipPipelineTrigger trigger, ClipItem originalClip)
    {
        Trigger = trigger;
        OriginalClip = originalClip;
        CurrentText = originalClip.PlainText;
    }
}

public interface IClipPipelinePlugin
{
    string Id { get; }
    string Name { get; }

    Task<bool> CanProcessAsync(ClipPipelineContext context);
    Task ProcessAsync(ClipPipelineContext context, CancellationToken cancellationToken);
}
